ofrak_ai/ofrak_ai/chatgpt.py

```diff
@@ -0,0 +1,96 @@
+import os
+import openai
+
+from dataclasses import dataclass
+from enum import Enum
+from openai.openai_object import OpenAIObject
+from openai.error import OpenAIError
+from typing import List, Dict, Optional
+
+from ofrak.model.component_model import ComponentConfig
+from ofrak.model.resource_model import ResourceAttributes
+from ofrak_ai.exponential_backoff import retry_with_exponential_backoff
+
+
+class ModelType(str, Enum):
+    """
+    Models ending in 4-digit long sequences are snapshots of their respective models. These are
+    static versions that receive no updates and will be deprecated 3 months after the newest
+    version is released.
+
+    GPT-4-32k models have a max token value of 32,768 (4x higher than standard GPT-4).
+
+    GPT-4 offers a larger model better optimized for chat than GPT-3.5-turbo as well as a higher max
+    token size (8,192 tokens vs 4,096). GPT-3.5 offers better cost efficiency.
+
+    Note: GPT-4 is currently in limited beta. Make sure that you have access to GPT-4 through
+    OpenAI before attempting to use any of the GPT-4 models with OFRAK AI.
+
+    For more information, visit https://platform.openai.com/docs/models/overview.
+    """
+
+    THREE_FIVE_TURBO = "gpt-3.5-turbo"
+    THREE_FIVE_TURBO_0301 = "gpt-3.5-turbo-0301"
+    FOUR = "gpt-4"
+    FOUR_0314 = "gpt-4-0314"
+    FOUR_32K = "gpt-4-32k"
+    FOUR_32K_0314 = "gpt-4-32k-0314"
+
+
+@dataclass
+class ChatGPTConfig(ComponentConfig):
+    """
+    :param api_key: the OpenAI API key to use
+    :param api_organization: the OpenAI API organization to use
+    :param model: the OpenAI model to use
+    :param system_message: a message which can be prepended to the conversation in API calls, which
+        is used for providing additional information to ChatGPT for generating responses
+    :param temperature: a measure of the randomness of ChatGPT's responses, between 0 (low) and
+        2 (high)
+    """
+
+    api_key: Optional[str] = os.getenv("OPENAI_API_KEY")
+    api_organization: Optional[str] = os.getenv("OPENAI_ORGANIZATION")
+    model: str = ModelType.THREE_FIVE_TURBO
+    system_message: Optional[str] = None
+    temperature: float = 1.0
+
+
+@dataclass
+class ChatGPTAnalysis(ResourceAttributes):
```

Comment:
@EdwardLarson  this is used by the program and function analyzers that I haven't pushed yet. I can remove it from this PR. 

If I understand you correctly about the specific attribute types, I think that may be difficult to accomplish given the probabilistic nature of chatgpt's output. In many cases it will use a similar structure between responses, but (to varying degrees of frequency) it can and will break that structure. That's why at the moment I'm treating responses as one text blob rather than parsing them into pieces and assigning to specific attributes.